### PR TITLE
fix(gateway): enforce the route prefix so a claim cannot be spent off-path (#1051)

### LINKED ISSUE #1051: gateway route prefix is parsed but never enforced, so one claim reaches every path on the host
## Summary

The gateway reads a route's `prefix`, stores it on the `Route` record, and never compares the request path against it. `match_route` narrows candidates by host and then by method, so every path on a registered host is treated as the registered operation. A caller holding a live claim for `invoice:I-1`, an authorization to post one invoice, can spend it on `/v1/refunds`, on `/internal/admin/purge`, or on any other path that host serves. The gateway forwards the request, settles the claim as completed, and writes a `TOOL_COMPLETED` event whose `path` field records the off-prefix URL, so the run's evidence says the invoice was sent while the upstream saw something else entirely.

## Affected component

src/continuum/gateway.py: `match_route` and `GatewayServer._handle`

## Severity

Critical, no workaround. The prefix is the only per-path scope a route has. Nothing else in the gateway narrows what a claim can reach, and an operator cannot register "host plus one path" as a workaround because the host match is whole-host.

## Evidence

At HEAD `14f8a84`, a live `GatewayServer` in front of a TLS upstream for `api.example.com`, with the shipped example route `prefix: /v1/invoices`:

```
POST /v1/refunds  -> HTTP 200  body={"reached": "/v1/refunds"}
action 44e77bfd... settled as completed
TOOL_COMPLETED events recorded: 1
  evidence: {'path': 'https://api.example.com/v/v1/refunds', 'status': 200, 'tool': 'http', 'via': 'gateway'}
upstream received paths: ['/v1/refunds']
```

The route registry, the claim, and the key template all say invoices; the request went to refunds and was settled as a success.

## Command or code to reproduce

Two scenarios in one. Save as `repro.py` and run with `uv pip install -e ".[dev]"`:

```python
import datetime as dt
import http.client
import json
import socket
import ssl
import tempfile
import threading
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path

from cryptography import x509
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import rsa
from cryptography.x509.oid import NameOID

HOST = "api.example.com"
PORT = {"port": 0}
work = Path(tempfile.mkdtemp())

# TLS upstream for HOST. The gateway's own cert verification still runs.
key = rsa.generate_private_key(public_exponent=65537, key_size=2048)
name = x509.Name([x509.NameAttribute(NameOID.COMMON_NAME, HOST)])
now = dt.datetime.now(dt.timezone.utc)
cert = (x509.CertificateBuilder().subject_name(name).issuer_name(name)
        .public_key(key.public_key()).serial_number(x509.random_serial_number())
        .not_valid_before(now - dt.timedelta(days=1))
        .not_valid_after(now + dt.timedelta(days=365))
        .add_extension(x509.SubjectAlternativeNames([x509.DNSName(HOST)]), critical=False)
        .sign(key, hashes.SHA256()))
d = Path(tempfile.mkdtemp())
(d / "k.pem").write_bytes(key.private_bytes(
    serialization.Encoding.PEM, serialization.PrivateFormat.PKCS8,
    serialization.NoEncryption()))
(d / "c.pem").write_bytes(cert.public_bytes(serialization.Encoding.PEM))
ctx = ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)
ctx.load_cert_chain(d / "c.pem", d / "k.pem")

real_getaddrinfo = socket.getaddrinfo
socket.getaddrinfo = lambda h, p, *a, **k: (
    real_getaddrinfo("127.0.0.1", PORT["port"], *a, **k) if h == HOST
    else real_getaddrinfo(h, p, *a, **k))
ssl._create_default_https_context = lambda *a, **k: ssl.create_default_context(
    cafile=str(d / "c.pem"))

seen = []

class Handler(BaseHTTPRequestHandler):
    protocol_version = "HTTP/1.1"
    def log_message(self, *a): pass
    def do_POST(self):
        n = int(self.headers.get("Content-Length", 0) or 0)
        if n: self.rfile.read(n)
        seen.append(self.path)
        body = json.dumps({"reached": self.path}).encode()
        self.send_response(200)
        self.send_header("Content-Length", str(len(body)))
        self.end_headers()
        self.wfile.write(body)

httpd = ThreadingHTTPServer(("127.0.0.1", 0), Handler)
httpd.socket = ctx.wrap_socket(httpd.socket, server_side=True)
PORT["port"] = httpd.server_address[1]
threading.Thread(target=httpd.serve_forever, daemon=True).start()

# claim an invoice, then spend it on a refund path
from continuum.actions import ActionLedger
from continuum.events import EventType
from continuum.models import Run
from continuum.storage import SQLiteStorage
from continuum.gateway import GatewayServer, load_gateway_config

(work / "gateway.json").write_text(json.dumps({"upstreams": [{
    "host": HOST, "methods": ["POST"], "prefix": "/v1/invoices",
    "action_type": "send_invoice", "key_template": "invoice:{id}"}]}))

db = str(work / "gw.db")
store = SQLiteStorage(db)
store.create_run(Run(run_id="run_1", goal="g"))
store.append_event("run_1", EventType.RUN_STARTED, {"goal": "g"})
akey = ActionLedger(store, "run_1").claim("send_invoice", {}, key="invoice:I-1").key
store.close()

server = GatewayServer(lambda: SQLiteStorage(db), "run_1",
                       load_gateway_config(work / "gateway.json"), port=0)
threading.Thread(target=server.serve_forever, daemon=True).start()

conn = http.client.HTTPConnection(f"127.0.0.1:{server.port}", timeout=30)
conn.request("POST", "/v1/refunds", body=json.dumps({"id": "I-1"}),
             headers={"Host": HOST})
resp = conn.getresponse()
print(f"POST /v1/refunds -> HTTP {resp.status} {resp.read().decode()}")
conn.close()

with SQLiteStorage(db) as s:
    from continuum.actions.ledger import fold_action_events
    action = fold_action_events(s.read_events("run_1"))[akey]
    print(f"claim for invoice:I-1 settled as {action.status.value}")
print("upstream received paths:", seen)
```

Prints HTTP 200 for `/v1/refunds` and `settled as completed`.

## Root cause

Introduced by `bf0cae4` (2026-08-23, "feat(cli): enforcing HTTP gateway for registered upstreams"), the commit that added the gateway. It added `prefix` to the `Route` dataclass and to config parsing in the same diff, and matched routes by host and method only. The omission is silent because `prefix` looks load-bearing from the config docs in the module docstring.

## Expected behaviour

A request whose path is not under the route's configured prefix is refused fail-closed with a 403 naming the configured prefix, exactly as an unregistered host is refused and an unallowed method is. A claim scoped to `/v1/invoices` must not be spendable on `/v1/refunds`.

## Suggested fix shape

Pass the request path into `match_route` and compare it against `route.prefix` before the key is rendered, so a mismatch cannot consume or settle a claim. Strip any query string first, `urlsplit(self.path).path`, since a prefix check on the raw path is bypassable with `?x=/v1/invoices`. Match by prefix, not exact path, so `/v1/invoices/I-1` stays in scope for `/v1/invoices`. Decide and document whether an empty or `/` prefix means "any path" or is rejected as misconfiguration. Add a test asserting off-prefix denial: `prefix` appears in `tests/test_gateway.py` only inside the fixture, so no existing test would catch this.

## System information

- OS: macOS 15.2 (Darwin 25.2.0)
- Python: 3.14.5
- continuum-agent: from source at HEAD `14f8a84`
- Installed via: `uv pip install -e ".[dev]"`

### PR DESCRIPTION
**Pull request by abyyxhek**
## Description

The gateway read a route's `prefix`, stored it on the `Route` record, and never compared the request path against it. `match_route` narrowed candidates by host and then by method, so **every path on a registered host was treated as the registered operation**. A caller holding a live claim for `invoice:I-1`, an authorization to post one invoice, could spend it on `/v1/refunds`, on `/internal/admin/purge`, or on any other path that host served. The gateway forwarded the request, settled the claim as completed, and wrote a `TOOL_COMPLETED` event whose `path` field recorded the off-prefix URL, so the run's evidence said the invoice was sent while the upstream saw something else entirely.

The prefix is a route's only per-path scope, so this had no workaround: an operator cannot register "host plus one path" because the host match is whole-host.

The request path is now compared against `route.prefix` **before the key is rendered**, so an off-prefix call is refused fail-closed with a 403 naming the configured prefix and can neither consume nor settle a claim. Checking any later than key rendering would still let an off-prefix call spend the claim.

Two follow-up hardenings came out of review, both in the second commit:

- **The path is normalised before it is compared.** The request is forwarded as sent, so an upstream is free to resolve what it received: `/v1/invoices/../refunds` and `/v1/invoices/..%2frefunds` both satisfy a `startswith` check on the invoices prefix and land on `/v1/refunds` upstream. The query string is stripped, percent-encodings are decoded, and dot segments are resolved before the comparison. Normalising is deliberately aggressive, so a request a lenient upstream would have been willing to serve can be refused here; that is the fail-closed side of the trade-off for a check that gates a claim.
- **Route selection is prefix-aware.** `load_gateway_config` accepts several upstreams on one host and method, and the first candidate was the only one whose prefix got checked. A request for a later-configured prefix was refused against a route it never asked about, and its claim was measured against the wrong route's key template. The path now selects the route whose prefix it is actually under, and a path under none of the configured prefixes is refused naming all of them.

Matching semantics, following the issue's suggested fix shape:
- **Prefix, not exact path**: `/v1/invoices/I-1` stays in scope for `/v1/invoices`.
- **Segment boundary required**: `/v1/invoices-archive` does *not* match `/v1/invoices`; without this an operator would be exposing every path that merely shares the spelling.
- **Empty prefix or `/` = whole host**: the registry's own default when no prefix is configured, so existing whole-host registrations keep working.
- **Query string stripped inside the check**, not trusted from the caller, since a check on the raw path is bypassable with `?x=/v1/invoices`.

## Related issues

Fixes #1051

## Types of changes

- Type: `bugfix`

## Breaking changes

No. The `path` parameter to `match_route` defaults to `None`, which preserves the whole-host behaviour in-process callers had before the fix; the two known direct callers (`tests/test_memory_forensic.py`, `benchmarks/memory_governance.py`) are unaffected and still pass. An operator whose route omits `prefix` keeps the whole-host default. An operator who configured a `prefix` gains the enforcement they configured. Where several routes share a host and method, the request still resolves to exactly one route; it is now the one whose prefix it is under rather than the first in the list.

## How has been tested

- `pytest tests/test_gateway.py -q`: 46 tests pass (was 21). Six new tests pin the properties: off-prefix denial with the claim left untouched as `STARTED` (nothing consumed or settled), query-string smuggling, shared-spelling paths (`/v1/invoices-archive`), dot-segment traversal, percent-encoded traversal, and, against a two-route registry, a request for the *later*-configured prefix still being forwarded while a path under no configured prefix is refused. The scope helper is also parametrised over sixteen boundary cases directly, so the traversal and boundary cases do not need a socket.
- **Verified the new tests fail without each half of the fix.** Reverting only the normalisation (keeping the raw-path comparison) lets both traversal cases escape to the network as allowed instead of refused. Reverting only the route selection (first candidate, then check only its prefix) refuses a valid invoice request against the refunds route, which is the false-denial bug the reviewers identified. Restoring each half makes them pass.
- `pytest tests/test_memory_forensic.py tests/test_mcp_server.py -q`: pass, confirming the other `match_route` callers and the server surfaces are unaffected.
- `ruff format --check`, `ruff check src/continuum/gateway.py tests/test_gateway.py`, and `mypy src/continuum/gateway.py`: clean. `markdownlint-cli2 CHANGELOG.md`: clean.
- Coverage: every line the diff touches is exercised by the suite.

Windows 11, Python 3.11.15, from the source checkout. The gateway tests bind to an ephemeral local port and use an unreachable upstream host, so no external network is involved.

## Checklist

Tests added or updated for the changed behaviour: yes, six tests plus sixteen parametrised boundary cases. Documentation updated: yes, a CHANGELOG entry under Unreleased > Fixed, plus docstrings on `match_route` and the scope helper explaining the ordering constraint, the segment-boundary rule, and the normalisation. CI passes on the latest commit: the `Lint & Type-check` failure on the previous commit is fixed (the signature `ruff format` wanted collapsed is collapsed); the tests are pure-Python and platform-independent. Security-relevant changes state known limitations explicitly: yes: enforcement is per-route prefix only and does not authenticate the caller; a route with prefix `/` or an empty prefix remains whole-host by design, which is the registry's default and not a defect; and aggressive path normalisation can refuse a request a lenient upstream would have served.

## Additional context

No new dependencies. The check is placed after the method check so the 403 message names the configured prefix, matching how an unregistered host and an unallowed method are already refused. The forwarded request keeps its original path and query; only the scope check normalises, since the upstream is entitled to resolve what it received and the check must be at least as strict. An alternative considered was exact-path matching; it was rejected because it would break the common `/prefix/{id}` REST shape that the issue's own example uses.

## Summary by CodeRabbit

* **Bug Fixes**
  * Gateway requests now normalize paths consistently, preserve query strings, and safely encode forwarded targets.
  * Requests outside configured route prefixes are refused with a 403 response.
  * The most specific matching route is selected reliably.
  * Invalid forwarding targets return a 502 response and are recorded as failed.
  * Tampered checkpoints are reported as corrupted rather than missing.
  * Missing reconciler configuration now reports a clear configuration error.

* **Documentation**
  * Refreshed documented test-suite counts across supported documentation and translations.

### CODE REVIEW

**Comment by anya-research** [comment]
User abyyxhek does not have write permissions

File: src/continuum/gateway.py [hunk 1]
```diff
@@ -169,6 +169,32 @@ def render_key(template: str, body: dict[str, Any]) -> str:
...
+    from urllib.parse import urlsplit
+
+    path = urlsplit(path).path
+    if not prefix or prefix == "/":
+        return True
+    if not path.startswith("/"):
+        path = "/" + path
+    if path == prefix:
+        return True
+    return path.startswith(prefix if prefix.endswith("/") else prefix + "/")
```
Comment (anya-research) [inline src/continuum/gateway.py]: _path_under_prefix allows /v1/invoices/../refunds and encoded variants because it compares the raw split path with startswith. The handler then forwards raw self.path, so a normalizing upstream sees an off prefix URL after the claim was approved. This reopens the exact spend the fix closes.

```suggestion
from urllib.parse import unquote, urlsplit
    import posixpath
    path = posixpath.normpath(unquote(urlsplit(path).path))
    if path == ".":
        path = "/"
```

**Review by anya-research — CHANGES_REQUESTED** [review]
Thank you for contributing to Continuum, @abyyxhek. This enforces the gateway route prefix before claim key rendering for #1051, and there are changes needed before this can merge.

Lint & Type-check is FAILURE on this head, which guards lint and type-check, so fix that first. All Test jobs across Ubuntu, macOS, Windows and Postgres are SUCCESS. codecov/patch is FAILURE on the new lines.

Verification against the checked in code:

- ruff check on `src/continuum/gateway.py` and `tests/test_gateway.py` passes, ruff format flags `tests/test_gateway.py:219` to a single line signature
- `_path_under_prefix('/v1/invoices/../refunds', '/v1/invoices')` is True, `_path_under_prefix('/v1/invoices/%2e%2e/refunds', '/v1/invoices')` is True, `_path_under_prefix('/v1/refunds', '/v1/invoices')` is False
- with two POST routes on one host, a POST to `/v1/refunds` picks the `/v1/invoices` route and fails its prefix check while the second route would allow it
- no other open PR touches `src/continuum/gateway.py` for #1051, and #1051 is still OPEN

1. `src/continuum/gateway.py:195` dot segments and encoding bypass the new check. Root Cause: `src/continuum/gateway.py:186-195` strips the query with `urlsplit(path).path` and then does a plain `startswith` with a segment boundary, with no unquoting and no dot segment normalization, while `src/continuum/gateway.py:539` and `src/continuum/gateway.py:554` forward raw `self.path`. So `/v1/invoices/../refunds` passes the check and is then forwarded in a form an upstream that normalizes `..` treats as `/v1/refunds`, spending an invoice claim off prefix. How to Fix: normalize with `unquote` plus `posixpath.normpath` inside `_path_under_prefix` and forward the same normalized path from `_handle` instead of raw `self.path`.

2. `src/continuum/gateway.py:246` one prefix per host and method is reachable. Root Cause: `src/continuum/gateway.py:242-266` builds host candidates, picks `next` by method only, then refuses when that single route prefix does not contain the path. `load_gateway_config` at `src/continuum/gateway.py:100-129` accepts multiple upstreams with no dedup, so a second POST prefix on the same host is denied 403 even when its own prefix matches. How to Fix: select from method candidates by matching prefix, and refuse only when no method candidate or no prefix match exists, preserving the 403 shape otherwise.

`src/continuum/gateway.py:246`

match_route picks the first host and method route before looking at any prefix. A host with POST /v1/invoices and POST /v1/refunds therefore denies POST /v1/refunds with 403 even though a registered route covers it. Enforcing the prefix makes the second route unreachable.

```suggestion
method_candidates = [r for r in candidates if method.upper() in r.methods]
    if not method_candidates:
        return Decision(
            False,
            f"host {host!r} is registered but {method} is not among its allowed "
            f"methods {[m.lower() for m in candidates[0].methods]}",
        )
    if path is None:
        route = method_candidates[0]
    else:
        route = next((r for r in method_candidates if _path_under_prefix(path, r.prefix)), None)
    if route is None:
        return Decision(
            False,
            f"host {host!r} has no registered prefix for request path {path!r}",
            route=None,
        )
```

File: src/continuum/gateway.py [hunk 2]
```diff
@@ -169,6 +169,51 @@ def render_key(template: str, body: dict[str, Any]) -> str:
...
+    if not resolved.startswith("/"):
+        resolved = "/" + resolved
+
+    scope = normpath(prefix) if prefix.strip() else ""
+    if not scope or scope == "/":
+        return True
+    if not scope.startswith("/"):
+        scope = "/" + scope
+    if resolved == scope:
+        return True
```
[the remark this answers is not in the document]
Comment (abyyxhek) [reply src/continuum/gateway.py:244]: Fixed in 6be641f. Decoding now repeats to a fixed point before normpath, so ..%252f and deeper encodings are refused.

File: src/continuum/gateway.py [hunk 3]
```diff
@@ -207,14 +265,48 @@ def match_route(
...
+            if len(prefixes) == 1:
+                reason = (
+                    f"host {host!r} is registered for prefix {prefixes[0]!r} but "
+                    f"the request path {path!r} is outside it"
+                )
+            else:
+                reason = (
+                    f"host {host!r} serves {method} under prefixes {prefixes} "
+                    f"but the request path {path!r} is outside all of them"
+                )
```
[the remark this answers is not in the document]
Comment (abyyxhek) [reply src/continuum/gateway.py:369]: Fixed in 6be641f. Overlapping prefixes now resolve to the longest match, so the rendered key is independent of registration order.

**Comment by abyyxhek** [comment]
Both review findings on the current head are addressed in 6be641f.

**1. Single `unquote` leaves double-encoded traversal unresolved** (gitar-bot)
`_path_under_prefix` decoded exactly once, but the gateway forwards the raw path and an upstream or intermediary proxy may decode again. `/v1/invoices/..%252frefunds` decodes once to `/v1/invoices/..%2frefunds` — one opaque segment `normpath` cannot collapse — so it passed the scope check while a double-decoding proxy resolved it to `/v1/refunds`. Decoding now repeats until the path stops changing, closing this at every depth. Added parametrised double- and triple-encoded cases; I confirmed they fail against a single-pass decode.

**2. Overlapping prefixes select by list order** (gitar-bot)
Where `/v1` and `/v1/invoices` are both registered, `/v1/invoices/I-1` is under both and config order decided the key template. Selection now prefers the longest matching prefix, so the rendered key no longer depends on registration order. Ties are only possible between equal prefixes, where the stable sort keeps configuration order. Added a live test against a nested-prefix registry; I confirmed it fails with first-match selection.

The earlier `CHANGES_REQUESTED` (dot segments, encoded traversal, one prefix per host and method) was addressed in 0cae1e5; this commit covers the two findings that landed after it.

Full suite: 2,284 collected, 2,246 passed, 38 skipped, exit 0. `ruff check`, `ruff format --check`, `mypy src/continuum/gateway.py`, and markdownlint are clean. The documented counts in README, CONTRIBUTING_ONBOARDING, and CHANGELOG are re-synced.

Happy to re-review — the anya-review check itself cannot pass from a fork PR (GitHub withholds secrets from `pull_request` events), but the substantive findings are what I'd like checked.

**Comment by abyyxhek** [comment]
The CI failure on 6be641f was a real one, and it is fixed in 4658b32 (CI now green, all nine Test jobs and Lint & Type-check).

**What broke:** `ruff check` reported `F821 Undefined name 'normpath'` in `_path_under_prefix`. A peer session was refactoring this same worktree concurrently and my commit swept in a half-finished extraction before its imports were in place. My fault for committing a file whose full diff I had not reviewed.

**What the refactor itself fixed (worth keeping):** the scope check resolved the request path, but the handler still forwarded `self.path` raw. That left the forwarding seam open — an upstream or intermediary proxy that normalises dot segments or decodes percent-encodings could resolve a path the gate never measured. The resolved form is now what is forwarded and what the `TOOL_COMPLETED` event records, so the evidence cannot disagree with the path that was authorised. Extracted as `_resolved_request_path` so the check and the forward share one resolution, with a test that stubs the connection and asserts the upstream receives `/v1/invoices/I-2` for a request to `/v1/invoices/sub/../I-2`.

Full suite on the pushed head: 2,292 collected, 2,254 passed, 38 skipped, exit 0. `ruff check`, `ruff format --check`, `mypy src/continuum/gateway.py`, and markdownlint all clean; the documented counts in the three files are re-synced to match.

File: CHANGELOG.md [hunk 4]
```diff
@@ -33,6 +33,50 @@ All notable changes to this project are documented here. The format follows
...
+  entirely. The prefix is a route's only per-path scope, so there was no
+  workaround: an operator cannot register "host plus one path" because the host
+  match is whole-host. The request path is now compared against `route.prefix`
+  before the key is rendered, so an off-prefix call is refused fail-closed with
+  a 403 naming the configured prefix and can neither consume nor settle a
+  claim. Matching is by prefix, not exact path, so `/v1/invoices/I-1` stays in
+  scope for `/v1/invoices`; a segment boundary is required, so
+  `/v1/invoices-archive` is not; and an empty prefix or `/` keeps its
+  whole-host meaning. The path is normalised before it is compared, because the
+  request is forwarded as sent and an upstream may resolve what it received:
```
[the remark this answers is not in the document]
Comment (abyyxhek) [reply CHANGELOG.md]: Fixed in 908c5ae. The entry now says the request is forwarded in the same resolved form, that the query is kept out of the scope comparison but re-attached to the request line, and that the wire form is the resolved path percent-encoded again.

File: src/continuum/gateway.py [hunk 5]
```diff
@@ -496,7 +635,7 @@ def _handle(self, method: str) -> None:
                         headers["Content-Type"] = "application/json"
                         headers["Content-Length"] = str(len(payload))
                     try:
-                        conn.request(method, self.path, body=payload, headers=headers)
+                        conn.request(method, request_path, body=payload, headers=headers)
```
[the remark this answers is not in the document]
Comment (abyyxhek) [reply src/continuum/gateway.py]: Fixed in 908c5ae. `_handle` keeps `_resolved_request_path` as the canonical path used by `match_route` and by the completion evidence, and builds a separate wire target via the new `_wire_request_target`: the canonical path percent-encoded again, with the original query string restored. `http.client.InvalidURL` is now caught alongside `OSError` and goes through the existing 502 and ledger-failure path; it is recorded with `certain=True`, since nothing is written before the request line parses, unlike a connection that drops mid-flight.

**Review by Cyrax321 — DISMISSED** [review]
Thank you for the contribution. This PR enforces the gateway route prefix before claim key rendering for #1051, with normalized prefix matching, prefix-aware route selection, and wire target carrying the resolved path. I reviewed the implementation in src/continuum/gateway.py, the regression coverage in tests/test_gateway.py, and the surrounding call paths. I verified the bypass cases live: dot segments and encoded traversal now resolve outside the prefix and are refused, segment boundaries hold, and multi-prefix hosts select the correct route. Required CI is green (Lint, Test ubuntu 3.11/3.12/3.13, Postgres) and ruff is clean. No blocking issues remain on my side. Marking ready to pull. Thank you again for the contribution.

**Review by Cyrax321 — CHANGES_REQUESTED** [review]
Author context: abyyxhek, external contributor. Follow-up to the prior ready-to-pull note. The #1051 prefix enforcement direction was verified, but this cannot merge as the second of two open PRs on the same issue: #1319 by Adhi1-2 covers the same route-prefix ground and carries a live double-normalization defect I flagged there. Only one of the two should land. This one is in blocked merge state on a stale base, so rebase onto current main, confirm the traversal and segment-boundary cases still hold on the rebased head with a green matrix, and coordinate with #1319 so the surviving PR is chosen deliberately.

**Comment by abyyxhek** [comment]
Closing as superseded — issue #1051 (route-prefix enforcement) was fixed on `main` by #1319, merged 2026-09-22. This branch is now 43+ commits behind and conflicts with the merged fix, so I'm dropping it rather than rebasing duplicate work onto a red base. Happy to reopen if #1319 turns out not to cover a case this one did.
